analyzers/src/SonarAnalyzer.Common/Helpers/ProjectConfigReader.cs

```diff
@@ -0,0 +1,76 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2021 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Xml;
+using System.Xml.Serialization;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace SonarAnalyzer.Helpers
+{
+    /// <summary>
+    /// This class reads and encapsulates <see cref="ProjectConfig"/>, exposing only the configuration our analyzers need.
+    /// </summary>
+    internal class ProjectConfigReader
+    {
+        private readonly Lazy<ProjectConfig> projectConfig;
+
+        public string AnalysisConfigPath => ProjectConfigValue?.AnalysisConfigPath;
+        public string FilesToAnalyzePath => ProjectConfigValue?.FilesToAnalyzePath;
+        public string OutPath => ProjectConfigValue?.OutPath;
+        public string ProjectPath => ProjectConfigValue?.ProjectPath;
+        public ProjectType? ProjectType => ProjectConfigValue?.ProjectType;
+        public string TargetFramework => ProjectConfigValue?.TargetFramework;
+
+        private ProjectConfig ProjectConfigValue => projectConfig.Value;
+
+        public ProjectConfigReader(AnalyzerOptions options)
+        {
+            projectConfig = new Lazy<ProjectConfig>(() => ReadContent(options));
+        }
+
+        private static ProjectConfig ReadContent(AnalyzerOptions options)
+        {
+            var sonarProjectConfig = options.AdditionalFiles.FirstOrDefault(IsSonarProjectConfig);
+            if (sonarProjectConfig == null || !File.Exists(sonarProjectConfig.Path))
+            {
+                return null;
+            }
+            try
+            {
+                var ser = new XmlSerializer(typeof(ProjectConfig));
+                using var fs = File.Open(sonarProjectConfig.Path, FileMode.Open, FileAccess.Read, FileShare.Read);
+                return (ProjectConfig)ser.Deserialize(fs);
+            }
+            catch (Exception)
```

Comment:
I think we should fail hard with AD0001 warning in this case. 
We're only in batch mode and something seriously unexpected happened here. 

Something was configured (like the exclusions, or a test project) and we'll silently behave very differently in this case. That would be a source of a poor user experience and big support pain

Reply:
I agree, however it's the same fail-soft approach we have when parsing the `SonarLint.xml` file. In both cases, the behavior will be unexpected...

Reply:
`SonarLint.xml` is little different IMO, as users are expected to work with that file, SL in connected mode provides the file to the project and users are expected to mess things up sometimes. And it can just naturally go out of sync.

Scope of `SonarProjectConfig` is narrow and users are not supposed to mess with it.

Reply:
Yes, we might rethink how to improve `SonarLint.xml`-troubleshooting user experience later. For example:
`Sxxxx: SonarLint.xml should exist and be valid` :D As we will have the path to the `csproj`, it could be the first rule 😄

We might actually raise some issue in the SonarLint.xml file itself 🐶 🥫